src/core/mac/sub_mac.cpp

```diff
@@ -1029,6 +1029,99 @@ void SubMac::StartTimerAt(Time aStartTime, uint32_t aDelayUs)
 #endif
 }
 
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+void SubMac::RadioSample(void)
+{
+#if OPENTHREAD_CONFIG_MAC_FILTER_ENABLE
+    VerifyOrExit(!mRadioFilterEnabled, IgnoreError(Get<Radio>().Sleep()));
+#endif
+
+    SetState(kStateRadioSample);
+
+    if (!RadioSupportsReceiveTiming())
+    {
+        UpdateRadioSampleState();
+    }
+
+#if OPENTHREAD_CONFIG_MAC_FILTER_ENABLE
+exit:
+#endif
+    return;
+}
+
+bool SubMac::IsRadioSampleEnabled(void) const
+{
+    bool ret = false;
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    ret = IsCslEnabled();
+#endif
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+    ret = ret || mIsWedEnabled;
+#endif
+
+    return ret;
+}
+
+/*
+ * The radio state (receive/sleep) is determined by the request from both CSL and WED:
+ * <1> If both CSL and WED request to enter sleep state, the radio is set to sleep state.
+ * <2> If either CSL or WED requests to enter the receive state and the other requests to enter sleep state, the radio
+ *     is set to receive state using the channel that is requested to enter the receive state.
+ * <3> If both CSL and WED request to enter the receive state, the radio is set to the receive state using the CSL
+ *     channel.
+ *
+ * The diagram below illustrates how to set the radio state based on the request of WED and CSL.
+ *
+ * CSL   ------========------------========------------========------------========---
+ *             ^       ^
+ *             |       |
+ *             | mIsCslSampling=false
+ *     mIsCslSampling=true
+ *
+ * WED   -----------++++++++----------------++++++++----------------++++++++----------
+ *                  ^       ^
+ *                  |       |
+ *                  | mIsWedSampling=false
+ *         mIsWedSampling=true
+ *
+ * Radio ------========+++++-------========-++++++++---========-----+++++++========---
+ *             ^       ^    ^
+ *             |       |    |
+ *             |       | Radio::Sleep()
+ *             |  Radio::Receive(WedCh)
+ *      Radio::Receive(CslCh)
+ */
+void SubMac::UpdateRadioSampleState(void)
+{
+    VerifyOrExit(mState == kStateRadioSample);
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    if (mIsCslSampling)
+    {
+        IgnoreError(Get<Radio>().Receive(mCslChannel));
+    }
+    else
+#endif
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+        if (mIsWedSampling)
+    {
```

Comment:
Can we use `ExitNow()` to avoid the `else if` and associated odd formatting that comes with the pre-processor statements?